internal/test_unit/python/unit.py

```diff
@@ -0,0 +1,371 @@
+            test_result = "pass" if (ret_ok and out_ok) else "FAIL"
+            logger.info(f"\t- {test_result} - \t{round(t_elaps.total_seconds(),3)} sec")
+
+            #   # on failure, print the problematic test and exit, if requested
+            if not (ret_ok and out_ok):
+                logger.info("\n".join(set_envs) + cmd + cmd_outs + "\n".join(unset_envs) + "\n")
+                if not noexit:
+                    sys.exit(1)
+
+
+def build_tests(test_root):
+    """
+    Parse the test components.
+
+    Take an ElementTree element extracted from a unit test xml file.
+    Return a list of all tests, where each test is represented as a dictionary,
+    with its keys representing each test component.
+
+    Parameters
+    ----------
+    test_root : ElementTree element
+        parsed from XML file containing the unit test(s) to perform
+
+    Returns
+    -------
+    test_list: 
+        list of test dicts, containing test attributes parsed from xml object
+
+    """
+
+    # define logger
+    logger = logging.getLogger(__name__)
+
+    # find all tests in test_xml, and create a dictionary of attributes for each test
+    test_list = []
+    for test_el in test_root.iter('test'):
+        test = {}
+        try:
+            test['name'] = test_el.attrib['name']
+        except KeyError:
+            logger.error("ERROR: name attribute not found for test")
+            raise
+            # should just fail this one test and not fully exit?
+
+        for el in test_el:
+            if (el.tag=='exec' or el.tag=='param'):
+                test[el.tag] = repl_env(el.text)
+            elif el.tag=='output':
+                test['out_pnc'] = []
+                test['out_gnc'] = []
+                test['out_stat'] = []
+                test['out_ps'] = []
+                test['out_exist'] = []
+                test['out_not_exist'] = []
+                output_names = {
+                    'point_nc' : 'out_pnc',
+                    'grid_nc' : 'out_gnc',
+                    'stat' : 'out_stat',
+                    'ps' : 'out_ps',
+                    'exist' : 'out_exist',
+                    'not_exist' : 'out_not_exist',
+                }
+                for output_el in el:
+                    test[output_names[output_el.tag]].append(repl_env(output_el.text))
+
+            elif el.tag=='env':
+                env_dict = {}
+                for env_el in el:
+                    try:
+                        env_name = env_el.find('name').text
+                        env_dict[env_name] = env_el.find('value').text
+                        if not env_dict[env_name]:
+                            env_dict[env_name] = ''
+                    except AttributeError:
+                        logger.error(f"ERROR: env pair in test \\{test['name']}\\ missing name or value")
+                        raise
+                        # should just fail this one test and not fully exit?                      
```

Comment:
Similar to above, in my opinion it is good to fully fail here since it is a configuration issue introduced by the user.